### PR TITLE
Filter all other dependency lists with skipOptionalDependencies

### LINKED ISSUE #81: skipOptionalDependencies doesn't appear to do anything?
Building `chokidar` on Linux fails because `fsevents` is Darwin only. Adding `fsevents` to the `skipOptionalDependencies` doesn't change the behavior at all. Furthermore the code doesn't do anything with that binding except remove it from the final attribute set.

Am I misunderstanding how `skipOptionalDependencies` is intended to be used?

### PR DESCRIPTION
**Pull request by ixmatus**
Fixes #81

This was tested with the `chokidar` npm package which uses `fsevents`, a package that does not work on Linux and is an optional dependency.

Building `chokidar` fails with the following output (truncated for brevity):

```
...
npm ERR! code EBADPLATFORM
npm ERR! notsup Unsupported platform for fsevents@1.1.3: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
npm ERR! notsup Valid OS:    darwin
npm ERR! notsup Valid Arch:  any
npm ERR! notsup Actual OS:   linux
npm ERR! notsup Actual Arch: x64
npm verb exit [ 1, true ]

npm ERR! A complete log of this run can be found in:
npm ERR!
/tmp/nix-build-fsevents-1.1.3-nodejs-8.9.4.drv-0/package/.npm/_logs/2018-02-18T22_06_04_943Z-debug.log
...
```

Overriding `chokidar` and adding `fsevents` to `skipOptionalDependencies` like so:

```
...
    (nodeLib.generatePackages {
      nodePackagesPath = ./nodePackages;
    }).override {
      overrides = nodePackagesNew: nodePackagesOld: {
        chokidar = nodePackagesOld.chokidar.overrideNodePackage (oldAttrs: {
          skipOptionalDependencies = (oldAttrs.skipOptionalDependencies or []) ++ [
            "fsevents"
          ];
        });
      };
    }
```

... allows the build of `chokidar` to succeed:

```
...
building
npm install --registry=http://notaregistry.078d3fd5b0-chokidar-2.0.2.com --fetch-retry-mintimeout=0 --fetch-retry-maxtimeout=10 --fetch-retries=0 --userconfig=/dev/null --nodedir=/nix/store/mlycgcw8j8bwqjsazp7hr1axiqnwhc4y-node-sources --production
installing
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/masxiwniv9srsj7bs4p72m14gy2s1znk-chokidar-2.0.2-nodejs-8.9.4
patching script interpreter paths in /nix/store/masxiwniv9srsj7bs4p72m14gy2s1znk-chokidar-2.0.2-nodejs-8.9.4
checking for references to /tmp/nix-build-chokidar-2.0.2-nodejs-8.9.4.drv-0 in /nix/store/masxiwniv9srsj7bs4p72m14gy2s1znk-chokidar-2.0.2-nodejs-8.9.4...
/nix/store/masxiwniv9srsj7bs4p72m14gy2s1znk-chokidar-2.0.2-nodejs-8.9.4
```

### CODE REVIEW

File: nix-libs/nodeLib/buildNodePackage.nix [hunk 1]
```diff
@@ -22,6 +22,12 @@ let
                        attrValues concatStringsSep optionalString filter
                        optionalAttrs;
 
+  # Like `subtractLists` but tests each dependency element in the
+  # minuend list for membership in the subtrahend list using that
+  # dependency's `basicName`.
+  subtractDependencyLists = subtrahend:
+    filter (x: !(elem x.basicName subtrahend));
```
Comment (adnelson) [inline nix-libs/nodeLib/buildNodePackage.nix]: Could using `basicName` run into trouble with namespaced packages?

**Comment by adnelson** [comment]
Maybe I'm reading this wrong, but it looks like elements of `skipOptionalDependencies` are being used to skip *any* kind of dependency, not just optional ones. If this is the case, there still might be uses for it, but we might want its name to reflect that. Also, shouldn't such behavior be accomplished by overriding `deps` or `devDependencies` etc?

File: nix-libs/nodeLib/buildNodePackage.nix (hunk 1 above)
Comment (ixmatus) [reply nix-libs/nodeLib/buildNodePackage.nix]: Good question. I need to check that.

**Comment by ixmatus** [comment]
@adnelson that was the intention because as I understood it, an optional dependency _can_ be skipped by NPM and it takes precedence over other dependencies. I went to check my understanding and I think this is only true for the `dependencies` set and not `devDependencies` or `peerDependencies`:

> Entries in optionalDependencies will override entries of the same name in dependencies, so it's usually best to only put in one place.

This means that dependencies _can_ appear in both sets. Which is the case with the `chokidar` package.

Since we `npm install` individual packages, in this case, `fsevents` doesn't look like an optional dependency to npm which is why it fails.

Because of this behavior, I think we should filter `optionalDependencies` and `deps` but not the other ones.

What do you think?

File: nix-libs/nodeLib/buildNodePackage.nix (hunk 1 above)
Comment (ixmatus) [reply nix-libs/nodeLib/buildNodePackage.nix]: In fact, I don't like this very much, I think we should just use the full package and use `lib.subtractLists`...

**Comment by adnelson** [comment]
It's possible that the whole idea of optional dependencies just doesn't really make sense with nix. If a package's dependencies are failing, its definition should be modified to remove the problematic ones. Providing `skipOptionalDependencies` seems kind of hacky, since its behavior (and much more) can be accomplished with overrides. Is there a reason to keep it?

**Comment by ixmatus** [comment]
Yeah I agree with you. I think we should remove `skipOptionalDependencies` since it currently does nothing. You are also right that the correct way to deal with this, I think, is to do one of two things:

- if an optional dependency is _broken_ because it doesn't build on a specific platform, then the right thing to do is override the optional dependency and mark it broken for that specific platform (as would be the case for `fsevents` on Linux)
- if you simply want it excluded from the package's dependency set and it cannot be built in isolation for some reason, then override the package depending on it and filter it out of the relevant dependency lists

**Comment by ixmatus** [comment]
I'm closing this PR as a result of our conversation. I will open a new one to remove `skipOptionalDependencies`.

**Pull request closed without merging**